clusterloader2/pkg/measurement/common/loadbalancer_nodesync_latency.go

```diff
@@ -0,0 +1,280 @@
+func (s *LoadBalancerNodeSyncMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	s.client = config.ClusterFramework.GetClientSets().GetClient()
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	switch action {
+	case "measure":
+		if err := s.selector.Parse(config.Params); err != nil {
+			return nil, err
+		}
+		s.waitTimeout, err = util.GetDurationOrDefault(config.Params, "waitTimeout", defaultNodeSyncLatencyTimeout)
+		if err != nil {
+			return nil, err
+		}
+		return nil, s.measureNodeSyncLatency()
+	case "gather":
+		if err := s.labelNodeForLBs(false); err != nil {
+			return nil, err
+		}
+		return s.gather(config.Identifier)
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+	return nil, nil
+}
+
+func (s *LoadBalancerNodeSyncMeasurement) Dispose() {}
+
+func (s *LoadBalancerNodeSyncMeasurement) String() string {
+	return loadBalancerNodeSyncLatencyName + ": " + s.selector.String()
+}
+
+func (s *LoadBalancerNodeSyncMeasurement) measureNodeSyncLatency() error {
+	ctx := context.Background()
+	options := metav1.ListOptions{}
+	s.selector.ApplySelectors(&options)
+	svcList, err := s.client.CoreV1().Services(s.selector.Namespace).List(ctx, options)
+	if err != nil {
+		return err
+	}
+
+	// lbSvcMap contains key->service mapping for load balancer type services
+	lbSvcMap := map[string]v1.Service{}
+	for _, svc := range svcList.Items {
+		if svc.Spec.Type == v1.ServiceTypeLoadBalancer {
+			key, err := runtimeobjects.CreateMetaNamespaceKey(&svc)
+			if err != nil {
+				return fmt.Errorf("meta key created error: %v", err)
+			}
+
+			lbSvcMap[key] = svc
+		}
+	}
+	totalLbSvc := len(lbSvcMap)
+
+	// trigger node sync by picking a node and add exclude lb label
+	nodeList, err := s.client.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, node := range nodeList.Items {
+		if !isCandidateNode(node) {
+			continue
+		}
+		s.triggerNodeName = node.Name
+	}
+
+	if s.triggerNodeName == "" {
+		return fmt.Errorf("failed to find a node candidate to trigger nodesync from node list: %v", nodeList.Items)
+	}
+	if err = s.labelNodeForLBs(true); err != nil {
+		return err
+	}
+
+	now := time.Now()
+	for key := range lbSvcMap {
+		s.svcNodeSyncLatencyTracker.Set(key, phaseNodeSyncStart, now)
+	}
+
+	return wait.Poll(5*time.Second, s.waitTimeout, func() (bool, error) {
+		for key, svc := range lbSvcMap {
+			if _, found := s.svcNodeSyncLatencyTracker.Get(key, phaseNodeSyncCompelte); found {
+				continue
+			}
+
+			eventList, err := s.client.CoreV1().Events(s.selector.Namespace).Search(Scheme, &svc)
+			if err != nil {
+				klog.Errorf("failed to list events associated with svc %v: %v", svc, err)
+				continue
+			}
+
+			for _, event := range eventList.Items {
+				if event.Reason == nodeSyncEventReason {
+					s.svcNodeSyncLatencyTracker.Set(key, phaseNodeSyncCompelte, event.CreationTimestamp.Time)
+				}
```

Comment:
Should we break here? 

Also, will events be in reverse chronological order? (so the latest lb update shows up first?). Otherwise we might catch the event where the LB was updated for the first time right?

Reply:
Adding a break. 

The measurement only triggers one nodesync event, so there should be only one nodeSyncEvent occur.